### ISSUE TITLE
Generator enums

### ISSUE THREAD

**Issue by sylvia-leaf** [SFe-Team-was-taken/SFe#35]
# Generator enums

### Introduction

To add extra features such as round-robin sampling, we need to use unused generator enumerator values. However, there is insufficient research about how enum values are used.

Right now, work is being done to [reverse engineer SFCF](https://github.com/SFe-Team-was-taken/SFe/issues/7). We've already found out that SFCF uses unused generator enum values for DLS features. Once this reverse engineering process reveals all of the enum values used by SFCF, we can then fit in our extra features. 

This is the reason why the partial reverse engineering of SFCF (to find unusable enum values) should be completed before we attempt to fit in extra generator features.

---

### Category 1: high priority

- [ ] Round-robin sampling
    - Priority: The highest out of everything on this list
    - Difficulty: Low
    - ETA: Ver. 4.2
- [ ] More sample modes
    - Priority: High
    - Difficulty: Medium
    - ETA: Ver. 4.2

### Category 2: medium priority

- [ ] Exclusive class normal release
    - Priority: Medium
    - Difficulty: Low
    - ETA: Ver. 4.2
- [ ] Independent envelopes/LFOs for pitch/attenuation/filter
    - Priority: Medium
    - Difficulty: Medium
    - ETA: Ver. 4.2 or 4.3
- [ ] Filter types
    - Priority: Medium
    - Difficulty: Medium
    - ETA: Ver. 4.3
- [ ] Filter parameters
    - Priority: Medium
    - Difficulty: Medium
    - ETA: Ver. 4.3

### Category 3: low priority

- [ ] Rest of DLS features
    - Priority: Low
    - Difficulty: High
    - ETA: Ver. 4.4 to 4.10 
- [ ] Conditional start/key switch
    - Priority: Low
    - Difficulty: High (may be impossible with generators)
    - ETA: Ver. 4.7 or 5.0

### Category 4: not in SFe 4

- [ ] Subtractive synthesis
    - Priority: Low
    - Difficulty: Really high
    - ETA: Ver. 5.0
- [ ] FM synthesis
    - Priority: Low
    - Difficulty: Really really high
    - ETA: Ver. 5.1
- [ ] Multiple loops
    - Priority: Low
    - Difficulty: Likely to be impossible with generators
    - ETA: Ver. 6.0

---

### Other things to do

Mirrored from [here](https://github.com/SFe-Team-was-taken/SFe/issues/7):

- [ ] What happens if a generator beyond `60` is found?
    - [ ] If it works, what is the maximum that's allowed?
- [ ] How do legacy soundcards react to undefined (in SF2.04) generators?   
- [ ] Use spaces before `60` for category 1 features.
- [ ] Use spaces between `60` and `127` for category 2.
- [ ] Use spaces after `127` for category 3.

---

If you've got anything else to add to the list, please send a comment here.

**Comment by spessasus** [comment]
> What happens if a generator beyond 60 is found?
Where? I'm sure it differs on implementations. On some it may work, others might reject it. What then?

**Comment by spessasus** [comment]
> You actually have a good point. Maybe the question I asked was bad; I should've asked "are there any commonly used programs that error out on a generator value beyond 60?"

What counts as commonly used? I can only think of fluidsynth and BASSMIDI.

PS: Did you read the discussion I started in this repo?

**Comment by sylvia-leaf** [comment]
If FluidSynth and Bassmidi are the only frequently used players, then we should just test on these two? Even though Davy mentioned "FluidLight" in the original SF update proposal repository, I can't find a project on GitHub called that. And even if I could, it seems like it would be a fork of FluidSynth, right?

Also, talking about the conversation you made, you are right. We shouldn't care about legacy players. But the problem is that I worry that people will attempt to run SFe banks on their legacy player and then complain and blame many different people, whether that be whoever made the bank, or even the SFe team for doing something that doesn't work.

You say that the multiple different variants of SFe are problematic. Well, davy7125 said that it is ok to [delay 64-bit support](https://github.com/davy7125/polyphone/issues/205#issuecomment-2476017415) until it's truly ready with a good foundation, because the format would be incompatible anyway:

>Reading 64 bit indexes should be moved backward so that existing soundfont readers can read soundfont with the new format.

We can thus safely get rid of SFe64. This would eliminate one variant of SFe, leaving just two variants:
- the "kinda compatible" 32-bit SFe-static (SFe-S) variant (with its subset SFe-L)
- the incompatible 64-bit SFe-dynamic (SFe-D) variant

While this would still be two formats, SFe-S would eventually stop getting features (this was what I envisioned with SFe32/64 in the first place!). Eventually the only format left would be SFe-D. I think that this strikes the perfect balance between compatibility and feature potential. What do you think? 

The SFe-L subset would stay within the limit of generator `60`, and it would include the most important features that davy7125 stated (listed above). But I've been wanting to do more esoteric things with the format that uses **way** more generators than can be accommodated by the limit. But again, I really want SFe32 banks to load (with a quality reduction) on legacy players. But you say that even structurally compatible changes can break \*\*\*\* in the most weird ways. I don't know what to say at this point...

Sorry for sounding angry, and sorry if I annoyed you with how I was structuring the format. Thank you for your feedback.

**Comment by sylvia-leaf** [comment]
Ah, I get it now!

So you are wanting to rename the `sfbk` fourcc to `sfen`, so legacy players intentionally do **NOT** load the file? That's an interesting idea, and something that I'll almost certainly consider for the current or a future version of SFe! Sorry for not reading the proposal written in your discussion properly.

**Comment by sylvia-leaf** [comment]
All features in categories 1-3 are now being tracked in their own issue. Won't close this issue yet, as there are features in category 4 that need to be tracked.

**Comment by stgiga** [comment]
A friend of mine said these:

> I feel this might be too out of scope but maybe SFe could make you choose between linear and exponential ADSR, and have other LFOs than sine?
> So it can emulate the Beatnik soundbanks better
> Like, a square wave / saw / triangle LFO
> Also maybe a general line based envelope/LFO control like in Beatnik and OpenMPT?

Also support for Unisoc Mocor 8-bit A-law SF2 sample decoding was discussed.
